### ISSUE TITLE
alpha.24 regression: Non-dataclass inheriting from a dataclass breaks generics

### ISSUE THREAD

**Issue by MeGaGiGaGon** [astral-sh/ty#1427]
### Summary

This is a MRE of some code that stopped working when I tried `ty 0.0.1-alpha.24` [playground](https://play.ty.dev/d62bf40e-aa4a-432e-9e9a-2c4ab01d9c3c). If you add `@dataclass` to `ChildOfParentDataclass` the error goes away.
```py
# Gives incorrect error when using a dataclass parent:

from dataclasses import dataclass

@dataclass
class ParentDataclass[T]:
    value: T

# @dataclass  # Uncommenting this makes the error go away
class ChildOfParentDataclass[T](ParentDataclass[T]): ...

def uses_dataclass[T](x: T) -> ChildOfParentDataclass[T]:
    # Argument is incorrect: Expected `T@ChildOfParentDataclass`, found `T@uses_dataclass`
    return ChildOfParentDataclass(x)

# Works fine if the parent is a normal class

class Parent[T]:
    def __init__(self, value: T):
        self.value = value

class ChildOfParent[T](Parent[T]): ...

def uses_init[T](x: T) -> ChildOfParent[T]:
    return ChildOfParent(x)
```
Comparing alpha.23 to alpha.24:
```powershell
PS ~>uvx ty@0.0.1-alpha.23 check issue.py
Checking ------------------------------------------------------------ 1/1 files
All checks passed!
PS ~>uvx ty@0.0.1-alpha.24 check issue.py
Checking ------------------------------------------------------------ 1/1 files
error[invalid-argument-type]: Argument is incorrect
  --> issue.py:14:35
   |
12 | def uses_dataclass[T](x: T) -> ChildOfParentDataclass[T]:
13 |     # Argument is incorrect: Expected T@ChildOfParentDataclass, found T@uses_dataclass
14 |     return ChildOfParentDataclass(x)
   |                                   ^ Expected `T@ChildOfParentDataclass`, found `T@uses_dataclass`
15 |
16 | # Works fine if the parent is a normal class
   |
info: rule `invalid-argument-type` is enabled by default

Found 1 diagnostic
```

### Version

ty 0.0.1-alpha.24 (1fee7da8b 2025-10-23)

**Comment by sharkdp** [comment]
Thank you for reporting this. Something seems broken with the `__init__` methods of generic dataclasses:

```py
from dataclasses import dataclass

@dataclass
class ParentDataclass[T]:
    value: T

class ChildOfParentDataclass[T](ParentDataclass[T]): ...

# (self: ParentDataclass[Unknown], value: Unknown) -> None
reveal_type(ParentDataclass.__init__)

# (self: ParentDataclass[T@ChildOfParentDataclass], value: T@ChildOfParentDataclass) -> None
reveal_type(ChildOfParentDataclass.__init__)
```

**Comment by saada** [comment]
I've submitted a PR to fix this issue: https://github.com/astral-sh/ruff/pull/21159

The fix threads the child class's generic context through to the parent's synthesized `__init__` method, restoring the behavior from alpha.23. All tests passing and verified against mypy and pyright.
